crates/bevy_ecs/src/schedule/executor/multi_threaded.rs

```diff
@@ -322,9 +359,15 @@ impl<'scope, 'env: 'scope, 'sys> Context<'scope, 'env, 'sys> {
             let Ok(mut guard) = self.environment.executor.state.try_lock() else {
                 return;
             };
-            guard.tick(self);
+            let cached_system = guard.tick(self);
             // Make sure we drop the guard before checking system_completion.is_empty(), or we could lose events.
             drop(guard);
+            if let Some(cached_system) = cached_system {
+                // SAFETY: The prior executor run guarantees that this system execution will not alias any
+                // data stored in the world, that the system has the permission to run, and that the system's
+                // archetypes are updated.
+                unsafe { cached_system.run_system(self) };
+            }
             if self.environment.executor.system_completion.is_empty() {
```

Comment:
If other threads failed to lock the executor, running the task before looping will delay spawning dependent systems until the cached system is run.  

For example, if you have `add_systems(Update, ((a, b).chain(), (c, d).chain()))`, then `b` and `d` can run concurrently.  But if `c` completes while `a` holds the executor lock, then the thread that completed `a` will run `b` before running the executor to spawn `d`.  

You may want to do something like 

```rust
let mut cached_system = None;
loop {
    if let Ok(mut guard) = self.environment.executor.state.try_lock() {
        guard.tick(self, &mut cached_system);
        drop(guard);
        if !self.environment.executor.system_completion.is_empty() {
            continue;
        }
    }
    match cached_system.take() {
        Some(cached_system) => unsafe { cached_system.run_system(self) },
        None => break,
    }
}
```